Register CN of OpenVPN clients in DNS Resolver

znerol opened the pull request:
https://redmine.pfsense.org/issues/6847

pfSense-git-updates commented:
Before this pull request can be accepted you must first sign a CLA as described at https://www.pfsense.org/about-pfsense/#cla. Please read for more details.

jim-p commented on src/etc/inc/openvpn.inc:
```diff
@@ -719,6 +719,15 @@ function openvpn_reconfigure($mode, $settings) {
 				break;
 		}
 	}
+	if ($settings['dev_mode'] === 'tun' && isset($config['unbound']['enable']) && isset($config['unbound']['regovpnclients'])) {
```
Why lock this to mode 'tun'? It's also possible to make a tap mode OpenVPN without a bridge that has addresses managed by OpenVPN.

znerol replied:
According to [man openvpn](https://openvpn.net/index.php/open-source/documentation/manuals/65-openvpn-20x-manpage.html) the address script parameter ...

> can be an IPv4 address such as "198.162.10.14", an IPv4 subnet such as "198.162.10.0/24", or an ethernet MAC address (when **--dev tap** is being used) such as "00:FF:01:02:03:04".

The script cannot handle MAC addresses (hence we cannot operate in `tap` device mode) and neither IPv4 subnets. I believe we never get those when the `mode` config is one of `server_tls` or `server_tls_user` (please correct me if I'm wrong).

jim-p replied:
OK, that makes sense. Though you should then also reflect this limitation in the help/description of the GUI option.

jim-p commented on src/etc/inc/openvpn.inc:
```diff
@@ -719,6 +719,15 @@ function openvpn_reconfigure($mode, $settings) {
+			case 'server_tls':
+			case 'server_tls_user':
+				$domain = !empty($settings['dns_domain']) ? $settings['dns_domain'] : $config['system']['domain'];
+				$conf .= "learn-address \"/usr/local/sbin/openvpn.learn-address.sh $domain\"\n";
```
$domain should be wrapped in escapeshellarg()

jim-p commented on src/usr/local/sbin/openvpn.learn-address.sh:
```diff
@@ -0,0 +1,53 @@
+
+	add|update)
+		# Remove all configs which mention the FQDN
+		grep -l -null -F "local-data: \"$CN.$DOMAIN A $IP\"" $DIR/openvpn.client.*.conf | xargs -0 rm
```
Use braces around shell variables to reduce ambiguity, for example: `${CN}.${DOMAIN}` and so on. Many instances of this throughout the script.

jim-p commented on src/usr/local/sbin/openvpn.learn-address.sh:
```diff
@@ -0,0 +1,53 @@
+	add|update)
+		# Remove all configs which mention the FQDN
+		grep -l -null -F "local-data: \"$CN.$DOMAIN A $IP\"" $DIR/openvpn.client.*.conf | xargs -0 rm
+		rm -f "$CONF"
```
Needs a sanity check before doing rm based on a variable name alone.

jim-p commented on src/usr/local/sbin/openvpn.learn-address.sh:
```diff
@@ -0,0 +1,53 @@
+
+		chmod 644 "$TMPCONF" "$TMPSRV"
+		/usr/local/sbin/unbound-checkconf "$TMPSRV" && mv "$TMPCONF" "$CONF"
+		rm -f "$TMPCONF" "$TMPSRV"
```
Needs a sanity check before doing rm based on a variable name alone.

znerol replied:
What sort of sanity check do you have in mind here? `TMPCONF` and `TMPSRV` are generated using `mktemp` which I assume is the correct way to securely create temporary files with unique names.

jim-p replied:
At the very least, a test to make sure that both parameters have some value that (e.g. not empty) and that none of them are only '/'. Ideally, it should also test if the file exists before trying to use `rm -rf` on it rather than assuming it does. There are a couple other usages of rm in the script that could use the same treatment.

Basically, we need to make sure the script won't `rm -rf` the entire filesystem or more files than intended, especially if somehow the mktemp call does not return an expected result.

tl;dr: Make no assumptions, assume input is invalid until proven otherwise, especially when removing files using variable names.

znerol replied:
The script does not use `-r` flag to `rm`, there is no recursion. Also note that `$IP` and `$CONF` should be fine now due to f4e8b87

jim-p commented on src/usr/local/sbin/openvpn.learn-address.sh:
```diff
@@ -0,0 +1,53 @@
+
+	delete)
+		# CN is not set on delete.
+		rm -f "$CONF" && /bin/pkill -HUP -F "$PIDFILE"
```
Needs a sanity check before doing rm based on a variable name alone.

rbgarga commented:
I'll wait items pointed out by @jim-p to be addressed before start reviewing

jim-p commented:
Two more general suggestions that should be addressed:
1. Calls to executables (e.g. `grep`, `rm`) must use the full path to the utilities rather than relying on `$path`.
2. Variable references in shell scripts need braces in all references, e.g. `$IP` -> `${IP}`

znerol commented:
@jim-p even if a string only consists of exactly one variable? e.g. `/bin/pkill -HUP -F "$PIDFILE"` -> `/bin/pkill -HUP -F "${PIDFILE}"`

jim-p commented:
Yes, it is more clear and consistent with other code that way. Doing it half one way and half the other looks sloppy IMO.

znerol commented:
@jim-p I think I addressed all of the feedback now (thanks a lot, btw.). I did a smoke test on one of my hosts, and the seems to behave. Let me know if there is anything else I can do here.

jim-p commented on src/etc/inc/unbound.inc:
```diff
@@ -376,6 +376,9 @@ include: {$g['unbound_chroot_path']}{$cfgsubdir}/host_entries.conf
 # dhcp lease entries
 include: {$g['unbound_chroot_path']}{$cfgsubdir}/dhcpleases_entries.conf
 
+# OpenVPN client entries
```
Perhaps this should be wrapped in similar tests to the case in openvpn.inc, so it isn't looking for these files when they do not or cannot exist? Or at least test for `$config['unbound']['regovpnclients']`

jim-p commented on src/usr/local/sbin/openvpn.learn-address.sh:
```diff
@@ -0,0 +1,60 @@
+	case "${OP}" in
+
+		add|update)
+			TMPCONF=$(mktemp "${CONF}.XXXXXX")
```
Needs full path (`/usr/bin/mktemp`) on this and the following line

jim-p commented on src/usr/local/sbin/openvpn.learn-address.sh:
```diff
@@ -0,0 +1,60 @@
+DIR="/var/unbound"
+PIDFILE="/var/run/unbound.pid"
+
+if [ -n "${IP}" -a "$(basename ${IP})" = "${IP}" ]; then
```
Needs full path (`/usr/bin/basename`)
